### DISCUSSION TITLE
REST source not producing the expected schema

### DISCUSSION THREAD

**Discussion by spierepf** [Budibase/budibase#5035] [Help] [answered]
I've created a REST source in budibase. It uses the GET verb against

`https://boardgamegeek.com/xmlapi/search?search={{search}}`

The idea is that a user will type a search term into a text field, click a submit button and be presented with a list of boardgames from boardgamegeek that matches the search term. Each entry in the result will include the `objectid` and `name`.

To acheive that, I've implemented the following transformer:

```javascript
var retval = [];
for(let i = 0; i < data.boardgame.length; ++i) {
  var entry = {
    objectid: parseInt(data.boardgame[i]["$"]["objectid"]),
    name: data.boardgame[i]["name"]["_"]
  };
  retval.push(entry);
}
return [retval];
```

which gets me the JSON result that I expect:

```json
[
  {
    "objectid": 283819,
    "name": "Ancient Times (fan expansion for Dominion)"
  },
  {
    "objectid": 203184,
    "name": "Animals (fan expansion for Dominion)"
  },
  {
    "objectid": 69035,
    "name": "The Books of Magic (fan expansion for Dominion)"
  },
  {
    "objectid": 215435,
    "name": "Codex: Card-Time Strategy – Flagstone Dominion vs. Blackhand Scourge Expansion"
  },
  ...
}
```

However, when I click on the Schema tab I get a schema where the keys are integers and the value is always `Text`, while what I am expecting is two keys `objectid` and `name` with values of Number and Text respectively.

Can anyone tell me what I am doing wrong?

**Comment by melohagan** [comment, marked as answer]
Hi @spierepf 

The schema expects an object to be returned that represents each element. 

So simply remove the square brackets: `return retval;`

What is currently happening is that the array is being treated as the schema, so each index is a field.

**Reply by spierepf** [reply]
Unfortunately, when I remove the square brackets from `return [retval];` then I only get a single response:

```
{
  "objectid": 283819,
  "name": "Ancient Times (fan expansion for Dominion)"
}
```

What I am hoping for is to receive a list of BGG entries which match the search criteria. Is that possible?

**Reply by melohagan** [reply]
It should be okay that only one row is coming back in the datasource JSON. This is just for the purposes of the schema.

This is what my transformer looks like:

![Screenshot 2022-03-23 at 16 41 27](https://user-images.githubusercontent.com/101575380/159751157-8d44fc3b-2256-4520-b99a-4a7e794ddc9c.png)

Then within the linked table block, all the rows come back:

![Screenshot 2022-03-23 at 16 39 30](https://user-images.githubusercontent.com/101575380/159750780-3921a268-c83a-4f04-9cd7-1c470999c4b6.png)

**Reply by spierepf** [reply]
Ahh! Ok. I misunderstood how REST data source configuration screen worked. Thanks so much!
